### DISCUSSION TITLE
Simpler way for getting default value for Nullable<T>

### DISCUSSION THREAD

**Discussion by Giridharan21** [dotnet/csharplang#6764] [Language Ideas]
Hi, 

```
public class ClassA 
{
    public bool IsCompleted { get; set; }
    public bool IsEnabled { get; set; }
    public List<string> Names { get; set; }
}

```

I have faced some scenarios where I have to convert bool? to bool for if conditions, it is simpler for one or two conversions, if there are many, I have to repeat the same conversion (some conversions mentioned below). This is not a frequent scenario but it would be better if this can be removed or simplified. Please suggest if there is already a simpler way without repetition available for this scenario.

```
ClassA obj = null;
var isEnabled1 = obj?.IsEnabled == true;
var isEnabled2 = obj?.IsEnabled ?? default;
var isEnabled3 = obj?.IsEnabled ?? false; // or true
var isEnabled4 = Convert.ToBoolean(obj?.IsEnabled);
var isEnabled5 = (obj?.IsEnabled).GetValueOrDefault();
var isEnabled6 = obj?.Enabled is true; // From pinkfloydx33
```

```
if(obj?.IsEnabled == true && obj?.IsCompleted == true && obj?.Names?.Any(i=>i == "test") == true)
{
    // conditions met
}
if ((obj?.IsEnabled ?? false) && (obj?.IsCompleted ?? false) && (obj?.Names?.All(i =>!string.IsNullOrEmpty(i)) ?? false))
{
    // conditions met
}
```

```
// Suggestions
// Can logical operations support operations between Nullable<T> , T

if(obj?.IsEnabled && obj?.IsCompleted && obj?.Names?.All(i => !string.IsNullOrEmpty(i))) 
{
    // conditions met
}

//By default obj?.IsEnabled evaluates to (obj?.IsEnabled).GetValueOrDefault()
//Or someother syntax like obj??.IsEnabled which will automatically GetValueOrDefault

if (obj??.IsEnabled && obj??.IsCompleted && obj?.Names??.All(i => !string.IsNullOrEmpty(i)))
{
    // conditions met
}
```

**Comment by EgorBo** [comment]
@Giridharan21 I moved your suggestion to dotnet/csharplang repo since you seem to request changes to C# lang, correct me if I'm wrong.

**Comment by pinkfloydx33** [comment]
`obj?.Enabled is true` is another option. Or patterns in general:

```csharp 
// long form: "is true" throughout; nothing gained w/r/t "== true" 
if(obj?.IsEnabled is true && obj?.IsCompleted is true && obj?.Names?.All(i => !string.IsNullOrEmpty(i)) is true) 
{
    // conditions met
}

// tuple + pattern
if ((obj?.IsEnabled, obj?.IsCompleted, obj?.Names?.All(i => !string.IsNullOrEmpty(i))) is (true, true, true)) 
{
    // conditions met 
} 

// shortest (note obj/obj.Names are non-null post &&) 
if(obj is { IsEnabled: true, IsCompleted: true } && obj.Names.All(i => !string.IsNullOrEmpty(i))) 
{
    // conditions met
}

// shortest (same as above but introduces o variable, non-null) [from Richiban]
if(obj is { IsEnabled: true, IsCompleted: true } o && o.Names.All(i => !string.IsNullOrEmpty(i)))
{
    // conditions met
}

// short but *makes sure* Names is non-null (auto property!)
if(obj is { IsEnabled: true, IsCompleted: true, Names: {} n } && n.All(i => !string.IsNullOrEmpty(i))) 
{
    // conditions met
}
```

[SharpLab](https://sharplab.io/#v2:D4AQTAjAsAUCAMACEEB0AZAlgOwI4G5YFkIAWQuAZmTEQGFEBvWRV5agIwHsuAbRAJIBnOlwC2AB14BTAC7SAJk0QBzOfkRD1iAL4s2ITj37CAotgCGHGUsartW2Rr0w27RFiGyAPCngA+RAA5CzFpIWU1J01tHUQAXkRsAFdeXgBCCjd9VkNkUkQAWQAKBi4OACsASiYct0QAegbEXi5sFUQAMy4AJzEALkQAIkwI2R7k6SHEWQALHq5klVnF6OwuOZwOlQscRUQAdwaehtlh+MTxyem6t0xO4vKKgH5UM0trfdGZielEADJ/ognq9hKJJDJ5Epvlc/oDgZVXiEwkJXgBBNLFTAJQLpPxvIRBVK8ADyPVMklkAE8sVUajDfjVbmxmK56qwXOzWMzWE0ZskpH8ANSICQWWTyHrYHmIe6IYqPREE8xWGwAGgRLwJ4MFUI1INQyPC6Mx2PiuPxwiJaTJFIk1NpdNlEWKsI1bp+kydMtZXN0iBlMr5QhWPXkXnl63kmoaT0NoXCiAsPT+62wAFoUmlRVwI4CmWz2fdFRVncp3qrFIMPWDxLqq56/nF4XGjUJUBjeFicYg8RB4ATraTyZSaZg6QW/b6uZyuUHmiHeuGzsUhAmkxErFwAG5/DjJM44cZcBTJADGia4iG3ycwlY1aczxKqMuLTzLdgrnwU1d+GtrEJyA2sL+leLbxiiHamj2fYDlaxK2qOjovoW9TTuys7svOmihmc+5nAAVGIFgANaJkIyQpgRwQJhE3yPlm/DFBYB5XhICwSNIYZUukKF+m+lQfoIQgqt+v6TP+Ih1pCwF/jRKKDIwcTYP68LYFBXZmha/aDghI72mOE4BqhbjofUmH1IGqEuDoQA===) 

Edit: oops thought `Names` itself was nullable; examples corrected

**Reply by Richiban** [reply]
I'd add 

```cs
if(obj is { IsEnabled: true, IsCompleted: true } o && o.Names.All(i => !string.IsNullOrEmpty(i))) 
{
    // conditions met
}
```

to that list. To me the problem here is arising from repeatedly accessing the `obj` variable in a null-safe way, when it's much better to simply check the object for null once at the beginning and then access its properties without issue.

**Reply by pinkfloydx33** [reply]
I misinterpreted the OP (no coffee yet!) and thought `Names` was a `List<string>?`. It's not, so my third version can remove the `?` after `Names` and it should still be non-null due to the check on `obj` prior (no need for the extra `o` variable) also allowing removal of the final `is true`:

```csharp
if(obj is { IsEnabled: true, IsCompleted: true } && obj.Names.All(i => !string.IsNullOrEmpty(i))) { }
```

Of course that's also assuming (and in your version as well) that `Names` has been properly initialized. Given that it's an auto-property I personally wouldn't take that chance (even if it had a `required` since those can still be explicitly set to `null`) and thus prefer my final version that includes the null-check on `Names`--via the property pattern--with the `n` variable

**Reply by Giridharan21** [reply]
Hi @pinkfloydx33  & @Richiban, 
Thanks for your suggestions. 

```
ClassA obj = null;
```
I am assuming Names to be null. If I add another condition then I have to repeat the same conversion again ( == true or is true). 
Even in the shorter syntax, we have to mention explicitly whether it needs to be true or false. 
//{ IsEnabled: true, IsCompleted: true }

```
public class ClassB
{
  public List<int> Numbers {get;set;}
}
public class ClassA 
{
    public bool IsCompleted { get; set; }
    public bool IsEnabled { get; set; }
    public List<string> Names { get; set; }
    public ClassB classB {get;set;}
}
```
```
// From pinkfloydx33
if(obj?.IsEnabled is true && obj?.IsCompleted is true && obj?.Names?.All(i => !string.IsNullOrEmpty(i)) is true && obj?.classB?.Numbers?.All(i=> i!=0) is true)  
{
    // conditions met
}
```

Mostly I will be using multiple LINQ queries with different properties, like 
```
 obj?.Names?.All(i => !string.IsNullOrEmpty(i))
 ```
It would be better if don't need to mention 'is true'.

**Reply by pinkfloydx33** [reply]
Yah but even that can be merged into the pattern:

```csharp
if(obj is { IsEnabled: true, IsCompleted: true, Names: {} n, classB.Numbers: {} num } && n.All(i => !string.IsNullOrEmpty(i)) && num.All(i => !string.IsNullOrEmpty(i))) 
{
    // conditions met
}
```
Note the nested property pattern for numbers. 

And you can shorten it even more by creating a function that checks  !string.IsNullOrEmpty and calling it as a method group:

```csharp
if(obj is { IsEnabled: true, IsCompleted: true, Names: {} n, classB.Numbers: {} num } && n.All(NotNullOrEmpty) && num.All(NotNullOrEmpty)) 
{
    // conditions met
}

static bool NotNullOrEmpty(string s) 
   => !string.IsNullOrEmpty(s);
```

If you're doing it a lot, I'd say to find a way to Refactor it a bit more into reusable parts. Of course it doesn't compose as well when talking about `||` instead, but it should be possible to come up with something clean. 

All that aside, I agree with Svick's comments below. I wouldn't want a nullable boolean to behave like a boolean in these situations. I don't think adding a new syntax is worth it... especially one that differs by a single char (and when `??` already has meaning)

**Comment by svick** [comment]
C# very deliberately forces you to state every time whether a `null` `bool?` is supposed to mean `true` or `false` (and, as you've shown, has several succinct syntaxes to do it). There is no "default" and I doubt the language will ever introduce one, even in the form of something like `??.`.

Sure, there is `GetValueOrDefault()`, but I'd consider that more a technical detail of `Nullable<T>`, than logical part of a tri-state boolean.
